[Deps] Update Kotlin to 2.0.21

ParaskP7 opened the pull request:
Project Thread: paaHJt-7n2-p2
Associated Metrics: paaHJt-7n2-p2#comment-9682

This PR updates Kotlin to 2.0 ([2.0.21](https://href.li/?https://github.com/JetBrains/kotlin/releases/tag/v2.0.21)).

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

FYI: I recommend reviewing this PR commit-by-commit just to make sure that Kotlin has been updated appropriately and that there isn't any left-overs that need addressing.

PS: Main commits for some extra focus:
- [Build: Update jvm target & source/target compatibility to java 11](https://github.com/woocommerce/woocommerce-android/pull/12896/commits/04e5b15878940d6ba2a19ed6e4fe2a3dfe9c9ebc)
- [Analysis: Resolve inline property bf warnings on woocommerce module](https://github.com/woocommerce/woocommerce-android/pull/12896/commits/dc48d4a31482eab05cd3d6df56850e5828895756)
- [Analysis: Resolve non-public primary ctor warning on woocommerce module](https://github.com/woocommerce/woocommerce-android/pull/12896/commits/5d43e33501d2beb0c706b304a5203f4f2511d7ec)

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- Smoke test the app and verify everything is working as expected.
- Try to test for build time increases both, locally and on CI.
- You could also run the `./gradlew buildHealth` task locally and verify that everything is working as expected ([CI](https://buildkite.com/automattic/woocommerce-android/builds/25006)).
- Review the [project dependencies & build environment changes (via CI artifacts](https://buildkite.com/automattic/woocommerce-android/builds/24999#01930c72-f76f-423d-8618-cfa833d3ab1e) and assert correctness. (*)

(*) Unfortunately, [this comment](https://github.com/woocommerce/woocommerce-android/pull/12896#issuecomment-2465092770) exceeds 65400 characters and thus the only way for us to check on the diff is to do it via the CI artifacts. You could focus on:
- [tldr_diff_dependencies.txt](https://buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/61f6f123-531b-42f6-9d4d-55bbfa3b28ea/01930c72-da5d-481e-a20a-36e0fa2d6f82/01930c72-f76f-423d-8618-cfa833d3ab1e/build/reports/diff/tldr_diff_dependencies.txt?response-content-type=text%2Fplain&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQPCP3C7LWE5WLIO5%2F20241111%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241111T115916Z&X-Amz-Expires=600&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBcaCXVzLWVhc3QtMSJHMEUCIQDFSJ%2BfDC2B%2F3Bryob9XOoOHquUNcEPNedy67rva1llKgIgYllElHOlsPK9ZNCxD2jFOHJbhnVT9G1HU%2FahQ7MvKKYq%2BgMIoP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwwMzIzNzk3MDUzMDMiDBnp2Z%2F3YZGYHV4Q1yrOA45dyPUpJ7ZEJ8lAgKeMVyVSj3LgHVm8wgdrCkke5W4hSovfAHl0EsUPjR4q%2B3ann31We1G4jdtPQ6rGMwEkZ5qxm8zF728gwz7ietnYw0%2BoCtOapQJfT6%2FWs95Xgka2PjEZ1gahHB5wbVNXpFBePi%2B4N6j9iUKNO1jJEvj46n8xMfOzzK9keqySZtTWMJ9Xq%2Bn7%2FonJa4Z%2Be4fEAdTLLpOpFkCMuwvrOonAwxbGy2zW7%2BJBu1oVjgN7xmYT2wi0PaIE6NAmiIV%2Fe%2FsvdbP5CayLdnZoNtY2k%2FPIF2AmWdiMeSD7x9y%2Fi%2F3aCe5QjWgfrZPqG3r1jHt%2FG91bnmjb3rbbJfeGafvMOj0WqHkY5alze4Lv95NwUNTpxOe8tkQleEom%2BhWXC7V9TRYeK%2FwJygFlmwibptcG7tazxPEo%2FJjXmFuv5iBrIexNt7JE7mYYw2iWogpYxvzq2BQFmZlPKyySXVWzDUFxGmy0TmGzPkieFDBuH%2FATGaSk%2FY8CcQj3mfxm4CS%2FhBWBOoVHxrRSXWCJqNwJQbh0eT09%2BnZ7yAPW2ZAwQcI4GkwMKhBj66RpLMmHZeJFHUkSIeU%2FMtWM%2BAy1vU96qf%2FUfrNRJcZpdjCq1Ma5BjqlARRSndNn6mYrA9JvVRZVxDp%2Fub0i03zEDBKHn7nAPRX605ps8H8YjnpncIX7Ni0wJfig70Ii6qI%2B0Gu97IcLV2jbyke0Pn59auavefZnnl0gX8MG83xvSBPfSDpfuOBeY4Ozu0VpxQmxoOGZnt%2BxZUSHfuh%2Fb2LfjoGClbO1HIsJ5ZZw5Bc37dvTdtyzA%2FzCa47XRuOpEqxT40hER02d4PpADYPBnQ%3D%3D&X-Amz-SignedHeaders=host&X-Amz-Signature=02ff3c789717f173e13e62968d0c73f3232ff5c774770497282d06c11d919fac)
- [tldr_diff_build_env.txt](https://buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/61f6f123-531b-42f6-9d4d-55bbfa3b28ea/01930c72-da5d-481e-a20a-36e0fa2d6f82/01930c72-f76f-423d-8618-cfa833d3ab1e/build/reports/diff/tldr_diff_build_env.txt?response-content-type=text%2Fplain&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQPCP3C7LYUHOAH4A%2F20241111%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241111T115915Z&X-Amz-Expires=600&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBcaCXVzLWVhc3QtMSJHMEUCIQCVfBsTv49WNYw0Te3p0JTlpq0OH0H8HdPt62RBsBCeNgIgacyXzcx2izbpedRTShvT4fzFuzSFoqK8l6emroV5wPQq%2BgMIoP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwwMzIzNzk3MDUzMDMiDFicUrC2%2Fqntoi54lSrOA5qrk2ICP3VEG5Rhfp5vaqi8BULVgd3SeTaV4ceYDWRMlDHse4NF2Jq4PimM97yqSqVQRW61bblTnCTPjQ5eeH0bnzafh%2FflK1tVDC1LfwlavyDhMCNlMgqYKaCQ%2Bb3OEAryfbEPYJqJkZfP1%2BleKNPU88spFEznPfsWriQQpDTYBpRwhYxuiIKReZEOk4uh7AIX1gwZiO4t9aIY5cgtvs0%2FSs7JRXbpDSSm9IkxlVF0ABnU8lkP5hA3lQxuJKAtE6k%2F0uOt1LiEHwNg6qEUE6f9Iz3YOxNm6DVn%2FsEdM4fjyn2iQwqtuD6aC8XQ3LyezUjDZguLzx%2Fp6BXgML5yVr6sx6KZSLUFvO2ClJyarEaT%2BZ7hW5GsACS5%2BmLhF%2F77DcoLF8YKq14Dm8DOFwFjHUCNdJOTbJKlfBB%2F8nwPRBLSfPqwhpUv16zLOjPEQ2Dm1kglNC7S5r7R091gHE1DvRlVFF2XAW6R9d3CJsE4Xlc0vLWWSjE%2BOi798SRa4r%2Fsr8or8igrNIxNTm9uP%2BzjeKS8LGNg9Y%2FnhUvWCpO0mWSadc%2Fc2IQIO0v8fOHNEEdw6T2jhGXyUXjPG2F4eFzQmEYGYSfXlB2LkQ2pOWpO4jCj0sa5BjqlAQBf5s%2Bn%2FQebSEEAgFSIcY7UPX7OMdHZ%2BeKGeTQwIibX6os%2Bec3dxi%2FQcxJL%2F9GnMx%2FC2PckuErt0s%2FBGihwefaTqc8rUpbDhCmHX2D%2BaiHVnCySczwPz%2BL94DUU%2BKYPETWD6B8zELWQHXiQ25gUsTrzsyG%2BAi5y%2BQStuMxy9xNrlB3FFF4qRsNpqF8YSKAuhJJh1Q98Y4Qvc20X%2BYZT7idyK%2B2j5Q%3D%3D&X-Amz-SignedHeaders=host&X-Amz-Signature=9618209cca088e23fe79e676dbcbffd3c046a0ebda8941fdb950f6145038f5e3)

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Deps: Update kotlin to 2.0.21 (ksp to 2.0.21-1.0.27)

Release Notes:
- Kotlin: https://github.com/JetBrains/kotlin/releases/tag/v2.0.21
- KSP: https://github.com/google/ksp/releases/tag/2.0.21-1.0.27

ParaskP7 committed:
Build: Remove unnecessary kotlin options jvm target configuration

FYI: This kotlinOptions/jvmTarget configuration is already available on
the root level 'build.gradle' file, which applies that automatically to
all modules.

ParaskP7 committed:
Build: Update jvm target & source/target compatibility to java 11

Although this change is not required, it is high time it is done in
order to update the Java target and compatibility across the project and
thus be prepared for any future updates.

------------------------------------------------------------------------

FYI: For example, on JP/WPAndroid this change was actually necessary
because otherwise updating to Kotlin 2.0 was failing the build with the
below:

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:processors:compileKotlin'.
> Could not resolve all files for configuration
   ':libs:processors:compileClasspath'.
   > Could not resolve com.google.devtools.ksp:symbol-processing-api:
      2.0.21-1.0.26.
     Required by:
         project :libs:processors
      > Dependency resolution is looking for a library compatible with
       JVM runtime version 8, but 'com.google.devtools.ksp:
       symbol-processing-api:2.0.21-1.0.26' is only compatible with JVM
       runtime version 11 or newer.

PR Commit: https://github.com/wordpress-mobile/WordPress-Android/pull/
21428/commits/6abe01ec983d0765d4bd1466030097109867ac56

ParaskP7 committed:
Build: Migrate jvm target & src/target compatibility to version catalogs

With this change, changing the 'java' version value within the
'libs.versions.toml' next time, will be all it takes to update JVM to
point to '17' or beyond.

ParaskP7 committed:
Deps: Add compose compile gradle plugin

Starting in Kotlin 2.0, the Compose Compiler Gradle plugin is required
when compose is enabled. This change addresses that, along with removing
the outdated 'androidx-compose-compiler' configuration.

------------------------------------------------------------------------
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':WooCommerce'.
> com.android.builder.errors.EvalIssueException: Starting in Kotlin 2.0,
 the Compose Compiler Gradle plugin is required
  when compose is enabled. See the following link for more information:
  https://d.android.com/r/studio-ui/compose-compiler

ParaskP7 committed:
Analysis: Suppress deprecated in java warnings on woocommerce module

Warning Messages: "This declaration overrides a deprecated member but is
not marked as deprecated itself. Please add the '@Deprecated' annotation
or suppress the diagnostic."

ParaskP7 committed:
Analysis: Resolve inline property bf warnings on woocommerce module

Warning Messages: "Inline property cannot have a backing field.
This will become an error in Kotlin 2.1."

FYI: For both, the 'ProductListAdapter' and 'AttributeBaseAdapter'
classes, this 'inline' keyword got added to 'clickListener' and
'onItemClick' correspondingly, but for no good reason other than
fixing an unknown Lint error, and then, probably due to copy-pasting.

For more info refer to commits below:
- 499904abc1a055954e20be0f41d2157b2a6f065e
- 16fb4f52ef7d0fefeaaee49a8cf3ae1a8c0ae82e

ParaskP7 committed:
Analysis: Resolve non-public primary ctor warning on woocommerce module

Warning Messages: "Non-public primary constructor is exposed via the
generated 'copy()' method of the 'data' class."

FYI: There is no good reason for this constructor to be 'private',
that is, instead of it being the default 'public'.

ParaskP7 committed:
Test: Fix 6 test failures on kotlin no when branch matched exception

------------------------------------------------------------------------

UpdateAnalyticsHubStatsTest > when syncing stats data starts with
 forceUpdate false, then follow data store response FAILED
    kotlin.NoWhenBranchMatchedException at
     UpdateAnalyticsHubStatsTest.kt:478

UpdateAnalyticsHubStatsTest > when data store allows new stats fetch,
 then request data with ForceNew strategy FAILED
    kotlin.NoWhenBranchMatchedException at
     UpdateAnalyticsHubStatsTest.kt:385

UpdateAnalyticsHubStatsTest > when syncing stats data stats with Stored
 strategy, then do not store the timestamp FAILED
    kotlin.NoWhenBranchMatchedException at
     UpdateAnalyticsHubStatsTest.kt:526

UpdateAnalyticsHubStatsTest > when syncing stats data starts with
 forceUpdate true, then trigger update with ForceNew Strategy FAILED
    kotlin.NoWhenBranchMatchedException at
     UpdateAnalyticsHubStatsTest.kt:453

UpdateAnalyticsHubStatsTest > when selection type is CUSTOM, then follow
 the data store and request data with Stored strategy FAILED
    kotlin.NoWhenBranchMatchedException at
     UpdateAnalyticsHubStatsTest.kt:425

UpdateAnalyticsHubStatsTest > when data store does NOT allows net stats
 fetch, then request data with Saved strategy FAILED
    kotlin.NoWhenBranchMatchedException at
     UpdateAnalyticsHubStatsTest.kt:397

ParaskP7 committed:
Git: Add .kotlin directory to .gitignore

This is done in order to avoid having the generated during build time
Kotlin session files being seen by Git.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into deps/update-kotlin-to-2.0.21

wzieba commented on WooCommerce-Wear/build.gradle:
```diff
@@ -77,19 +79,13 @@ android {
-    }
-    kotlinOptions {
-        jvmTarget = '1.8'
+        sourceCompatibility JvmTarget.fromTarget(libs.versions.java.get()).target
```
I suggest simplifying this to:

```suggestion
        sourceCompatibility libs.versions.java.get()
```

ParaskP7 replied:
Good suggestion @wzieba ! 👍 

Frankly, I got/stole this idea from [PCAndroid](https://github.com/Automattic/pocket-casts-android/blob/1585aad387eb28883f61ba8d48224c81883e3080/build.gradle.kts#L129), but which is on Kotlin DSL for Gradle already. Thus, I thought it will be a good idea to use that now, versus just using `libs.versions.java.get()` directly. The only drawback being the boilerplate code with wrapping everything into `JvmTarget.fromTarget(...).target`. My thinking was:

1. **Type Safety**: The `JvmTarget.fromTarget()` method ensures that the provided version string is a valid JVM target. It will throw an exception if an invalid version is provided, catching potential errors early.
2. **Consistency with Kotlin DSL**: When using the Kotlin DSL for Gradle, JvmTarget integrates more naturally with the strongly-typed nature of Kotlin. `jvmTarget.set(javaTarget)` versus `jvmTarget.set(libs.versions.java.get())` which will not be allowed.

I don't feel strongly about it, so if you insist on changing that to `libs.versions.java.get()` after reading the above, I'll change it everywhere.

PS: Same applies to [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android), which already got merged into `trunk` with `JvmTarget.fromTarget(libs.versions.java.get()).target`. I was hoping this would have been discussed there instead so that we avoid having multiple ways of dealing with that per project. 😊

wzieba replied:
> Type Safety

I don't think type safety is important here: this is simply a version of Java, so if it's incorrect, then it'll be caught at the same as if you wrap it with `JvmTarget`. 

> Consistency with Kotlin DSL

Sorry, I'm not sure if I follow - as we don't use Kotlin DSL here, I don't think we should focus on Kotlin DSL.

It's a minor, so I don't insist. I just thought having less code would be simply better in this case.

> PS: Same applies to [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android), which already got merged into trunk with JvmTarget.fromTarget(libs.versions.java.get()).target. I was hoping this would have been discussed there instead

I can assure you I didn't do this on purpose! I simply spotted this improvement in this PR, not the other one.

ParaskP7 replied:
> I can assure you I didn't do this on purpose!

Wasn't implying anything! 🤣

> I don't think type safety is important here: this is simply a version of Java, so if it's incorrect, then it'll be caught at the same as if you wrap it with JvmTarget.

👍 

> Sorry, I'm not sure if I follow - as we don't use Kotlin DSL here, I don't think we should focus on Kotlin DSL.

Yea, I was trying to be smart about it and prepare in advance, not worth it (I guess). 👍 

> It's a minor, so I don't insist. I just thought having less code would be simply better in this case.

Let's have less code then! 😉 (yes, I know how you feel about 😉 -> 😜 🤣)

:onit:

ParaskP7 replied:
This is now done: a65218cb2ca83dcbc560912c1731e64d36a5d793

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/reviews/ReviewModeration.kt:
```diff
@@ -4,8 +4,7 @@ import com.woocommerce.android.model.ActionStatus
 
-@Suppress("DataClassPrivateConstructor")
-data class ReviewModerationRequest private constructor(
+data class ReviewModerationRequest(
```
> FYI: There is no good reason for this constructor to be 'private',
that is, instead of it being the default 'public'.

I think the reason was to disable possibility of providing `ReviewModarationRequest` with `timeOfRequest`. If we want to keep constructors signature of this class as it was but without having  private constructor, we have to perform a small refactor:

```kotlin
data class ReviewModerationRequest(
    val review: ProductReview,
    val newStatus: ProductReviewStatus,
) : Comparable<ReviewModerationRequest> {

    private val timeOfRequest = System.currentTimeMillis()

    override fun compareTo(other: ReviewModerationRequest): Int {
        return timeOfRequest.compareTo(other.timeOfRequest)
    }
}

```

ParaskP7 replied:
You're right @wzieba and thanks for this suggestion, a much better outcome, done: 8f94fc8324387f6b474a3a63c68b3ba223403fc3

wzieba reviewed (commented):
Leaving some suggestions. Is there any place to review build performance changes for this update?

ParaskP7 commented:
👋 @wzieba !

> Is there any place to review build performance changes for this update?

Don't you mean this: Associated Metrics: paaHJt-7n2-p2#comment-9682 (see PR description, on top) 🤔

Or, do you have something else in mind? 🤔

wzieba commented:
> Don't you mean this: Associated Metrics: paaHJt-7n2-p2#comment-9682 (see PR description, on top) 🤔

Yes, thanks!

ParaskP7 committed:
Refactor: Make time of request a private val on review moderation

This way both the non-public primary constructor being exposed via the
generated 'copy()' method of the 'data' class warning, as well as the
'DataClassPrivateConstructor' warning are solved, but at the same time
the signature and functionality of the 'ReviewModerationRequest' data
class remains unchanged.

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
12896#discussion_r1837889124

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into deps/update-kotlin-to-2.0.21

ParaskP7 committed:
Build: Replace jvm target from target with plain libs versions java get

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
12896#discussion_r1837878971

wzieba reviewed (approved):
Thanks for considering and applying improvements! All LGTM 👍

ParaskP7 commented:
👋 @samiuelson !

You being part of the product team, are you able to code review/test this Kotlin 2.0 change before us merging this to `trunk`. In addition to the 👍 from Wojtek, I would love if someone from the product team gives the final 👍 . Else, let me know if you're not available. I will then re-assign this to someone with more bandwidth. 🙏

samiuelson reviewed (approved):
Thanks for working on that, @ParaskP7. LGTM :shipit:

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into deps/update-kotlin-to-2.0.21

ParaskP7 commented:
Awesome @samiuelson , thanks for reviewing, testing and giving your final 👍 on this change, I updated this PR with the latest changed from `trunk`, will wait on CI to go 🟢 and then merge this! 🙇 ❤️ 🚀
